File: data_structure/Other/List2/list.cpp
#include <iostream>
#include "list.hpp"
using namespace std;

Node::Node(Data d){
    data = d;
    next = NULL;
    prev = NULL;
}

List::List(){
    first = NULL;
    last = NULL;
    listSize = 0;
}

void List::insert(Data data){
    insertEnd(data);
}

void List::insertEnd(Data data){
    Node* temp = new Node(data);
    if(listSize == 0){
        first = temp;
        last = temp;
    }else{
        Node* aux = last;
        aux->next = temp;
        temp->prev = aux;
        last = temp;
    }
    ++listSize;
}

void List::print(){
    Node* temp = first;
    while(temp != NULL){
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

List::~List(){
    cout << "Destructor Called!" << endl;
}

File: Workspace/Trabalho/src/structures/List.java
package structures;

public class List {
	private Node first;
	private Node last;
	private int size;
	
	public List() {
		this.first = null;
		this.last = null;
		this.size = 0;
	}
	
	public void insertEnd(int data) {
		insertPos(data, 0);
	}
	
	public void insertBegin(int data) {
		insertPos(data, size - 1);
	}
	
	public void insertPos(int data, int pos) {
		Node node = new Node(data);
		
		if(size == 0) {
			first = node;
			last = node;
		}
		else {
			if(pos == 0) {
				node.setNext(first.getNext());
				first = node;
			}
			else if(pos == (size - 1)) {
				last.setNext(node);
				last = node;
			}
			else {
				int i = 0;
				Node temp = first;
				Node prev = null;
				while(i < pos) {
					prev = temp;
					temp = temp.getNext();
					++i;
				}
				prev.setNext(node);
				node.setNext(temp);
			}
		}
		++size;
	}
	
	
}
